instrumentation/message-handler/library/src/main/java/io/opentelemetry/instrumentation/messagehandler/SqsBatchMessageHandler.java

```diff
@@ -0,0 +1,128 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.messagehandler;
+
+import com.amazonaws.services.lambda.runtime.events.SQSEvent;
+import io.opentelemetry.api.OpenTelemetry;
+import io.opentelemetry.api.trace.Span;
+import io.opentelemetry.api.trace.SpanBuilder;
+import io.opentelemetry.api.trace.SpanContext;
+import io.opentelemetry.api.trace.propagation.W3CTraceContextPropagator;
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.propagation.TextMapGetter;
+import io.opentelemetry.contrib.awsxray.propagator.AwsXrayPropagator;
+import io.opentelemetry.semconv.trace.attributes.SemanticAttributes;
+import java.util.Collections;
+import java.util.Locale;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class SqsBatchMessageHandler extends BatchMessageHandler<SQSEvent.SQSMessage> {
+  private static final Logger logger = Logger.getLogger(BatchMessageHandler.class.getName());
+
+  private static final String AWS_TRACE_HEADER_SQS_ATTRIBUTE_KEY = "AWSTraceHeader";
+  private static final String AWS_TRACE_HEADER_PROPAGATOR_KEY = "x-amzn-trace-id";
+
+  public SqsBatchMessageHandler(OpenTelemetry openTelemetry) {
+    super(openTelemetry);
+  }
+
+  public SqsBatchMessageHandler(OpenTelemetry openTelemetry, String messageOperation) {
+    super(openTelemetry, messageOperation);
+  }
+
+  public SqsBatchMessageHandler(
+      OpenTelemetry openTelemetry, String messageOperation, String spanName) {
+    super(openTelemetry, messageOperation, spanName);
+  }
+
+  @Override
+  protected void addMessagingAttributes(SpanBuilder spanBuilder) {
+    super.addMessagingAttributes(spanBuilder);
+    spanBuilder.setAttribute(SemanticAttributes.MESSAGING_SYSTEM, "AmazonSQS");
+  }
+
+  @Override
+  public SpanContext getParentSpanContext(SQSEvent.SQSMessage message) {
+    String parentHeader = message.getAttributes().get(AWS_TRACE_HEADER_SQS_ATTRIBUTE_KEY);
+
+    if (parentHeader == null) {
+      return null;
+    }
+
+    SpanContext spanContext = getParentSpanContextXRay(parentHeader);
+
+    if (spanContext != null) {
+      return spanContext;
+    }
+
+    spanContext = getParentSpanContextW3C(parentHeader);
+
+    if (spanContext != null) {
+      return spanContext;
+    }
+
+    logger.log(Level.WARNING, "Invalid upstream span context: {0}", parentHeader);
+    return null;
+  }
+
+  private static SpanContext getParentSpanContextW3C(String parentHeader) {
```

Comment:
I am hesitant adding this because SQS message does not have system attribute `traceparent`, it only appear if SQS publisher add custom system attribute by w3cPropagator.inject().

So, can we just keep getParentSpanContext() be public, for power users they can override this method for extracting w3c or b3 headers either in SQS system attirbute or normal message attribute?

Reply:
If the downstream service uses OTEL instead of X-Ray, then won't it use the W3C format?

Reply:
198 links checked.
  [✖] https://www.payara.fish/ → Status: 0
Error: Process completed with exit code 123.

This error below is unrelated to our code change.